cli/loader_lib.c

```diff
@@ -0,0 +1,194 @@
+    if (ret != 0) {
+        print_stderr("ERROR: _NSGetExecutablePath() failed\n");
+        exit(1);
+    }
+
+    if (realpath(nonreal_exe_path, exe_dir) == NULL) {
+        print_stderr("ERROR: realpath() failed\n");
+        exit(1);
+    }
+#elif defined(_OS_LINUX_)
+    // On Linux, we read from /proc/self/exe
+    int num_bytes = readlink("/proc/self/exe", exe_dir, PATH_MAX);
+    if (num_bytes == -1) {
+        print_stderr("ERROR: readlink(/proc/self/exe) failed\n");
+        exit(1);
+    }
+    exe_dir[num_bytes] = '\0';
+#elif defined(_OS_FREEBSD_)
+    // On FreeBSD, we use the KERN_PROC_PATHNAME sysctl:
+    int mib[4] = {CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, -1};
+    int exe_dir_len = PATH_MAX - 1;
+    int ret = sysctl(mib, 4, exe_dir, &exe_dir_len, NULL, 0);
+    if (ret) {
+        print_stderr("ERROR: sysctl(KERN_PROC_PATHNAME) failed\n");
+        exit(1);
+    }
+    exe_dir[exe_dir_len] = '\0';
+#endif
+    // Finally, convert to dirname
+    const char * new_dir = dirname(exe_dir);
+    if (new_dir != exe_dir) {
+        // On some plateforms, dirname() mutates.  On others, it does not.
+        memcpy(exe_dir, new_dir, strlen(new_dir)+1);
+    }
+    return exe_dir;
+}
+
+// Load libjulia and run the REPL with the given arguments (in UTF-8 format)
+int load_repl(const char * exe_dir, int argc, char * argv[])
+{
+    // Pre-load libraries that libjulia needs.
+    int deps_len = strlen(dep_libs);
+    char * curr_dep = &dep_libs[0];
+    while (1) {
+        // try to find next colon character, if we can't, escape out.
+        char * colon = strchr(curr_dep, ':');
+        if (colon == NULL)
+            break;
+
+        // Chop the string at the colon, load this library.
+        *colon = '\0';
+        load_library(curr_dep, exe_dir);
+
+        // Skip ahead to next dependency
+        curr_dep = colon + 1;
+    }
+
+    // Last dependency is `libjulia`, so load that and we're done with `dep_libs`!
+    void * libjulia = load_library(curr_dep, exe_dir);
+
+    // Next, if we're on Linux/FreeBSD, set up fast TLS.
+#if !defined(_OS_WINDOWS_) && !defined(_OS_DARWIN_)
+    void (*jl_set_ptls_states_getter)(void *) = dlsym(libjulia, "jl_set_ptls_states_getter");
+    if (jl_set_ptls_states_getter == NULL) {
+        print_stderr("ERROR: Cannot find jl_set_ptls_states_getter() function within libjulia!\n");
+        exit(1);
+    }
+    void * (*fptr)(void) = dlsym(NULL, "jl_get_ptls_states_static");
+    if (fptr == NULL) {
+        print_stderr("ERROR: Cannot find jl_get_ptls_states_static(), must define this symbol within calling executable!\n");
+        exit(1);
+    }
+    jl_set_ptls_states_getter((void *)fptr);
+#endif
+
+    // Load the repl entrypoint symbol and jump into it!
+    int (*entrypoint)(int, char **) = NULL;
+    #ifdef _OS_WINDOWS_
+        entrypoint = (int (*)(int, char **))GetProcAddress((HMODULE) libjulia, "repl_entrypoint");
```

Comment:
I wonder if it doesn't like how we've prevented initialization of the C++ runtime. What if instead you were to call the initialization routine from here:

```suggestion
    entrypoint = (int (*)(void))GetProcAddress((HMODULE) libjulia, "mainCRTStartup");
    return entrypoint();
```

Will `ld` even let us link the necessary code into libjulia.dll?

```
gcc crt2.o crtbegin.o $usual_arguments -lstdc++ -lmingw32 -lgcc_s -lgcc -lmoldname -lmingwex -lmsvcrt -lpthread -ladvapi32 -lshell32 -luser32 -lkernel32 -lmingw32 -lgcc_s -lgcc -lmoldname -lmingwex -lmsvcrt crtend.o
```

You can see this list of missing libraries by calling make with `LDFLAGS=-v`, and it'll print something like this:

 /usr/lib/gcc/i686-w64-mingw32/9.3-posix/collect2 -plugin /usr/lib/gcc/i686-w64-mingw32/9.3-posix/liblto_plugin.so -plugin-opt=/usr/lib/gcc/i686-w64-mingw32/9.3-posix/lto-wrapper -plugin-opt=-fresolution=/tmp/cc0tVtOg.res -plugin-opt=-pass-through=-lmingw32 -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lmoldname -plugin-opt=-pass-through=-lmingwex -plugin-opt=-pass-through=-lmsvcrt -plugin-opt=-pass-through=-lpthread -plugin-opt=-pass-through=-ladvapi32 -plugin-opt=-pass-through=-lshell32 -plugin-opt=-pass-through=-luser32 -plugin-opt=-pass-through=-lkernel32 -plugin-opt=-pass-through=-lmingw32 -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lmoldname -plugin-opt=-pass-through=-lmingwex -plugin-opt=-pass-through=-lmsvcrt -m i386pe -Bdynamic -o /data/vtjnash/julia/julia-win32/usr/bin/julia.exe /usr/lib/gcc/i686-w64-mingw32/9.3-posix/../../../../i686-w64-mingw32/lib/../lib/crt2u.o /usr/lib/gcc/i686-w64-mingw32/9.3-posix/crtbegin.o -L/data/vtjnash/julia/julia-win32/usr/lib/julia -L/data/vtjnash/julia/julia-win32/usr/lib -L/data/vtjnash/julia/julia-win32/usr/bin -L/usr/lib/gcc/i686-w64-mingw32/9.3-posix -L/usr/lib/gcc/i686-w64-mingw32/9.3-posix/../../../../i686-w64-mingw32/lib/../lib -L/usr/lib/gcc/i686-w64-mingw32/9.3-posix/../../../../i686-w64-mingw32/lib ./repl.o julia_res.o -ljulia --stack 8388608 --large-address-aware --no-whole-archive --export-all-symbols --version-script=/data/vtjnash/julia/src/julia.expmap --no-whole-archive -lpsapi -lkernel32 -lws2_32 -liphlpapi -lwinmm -ldbghelp -luserenv -lsecur32 -lssp --whole-archive /data/vtjnash/julia/julia-win32/usr/lib/libopenlibm.a --no-whole-archive -lstdc++ -lmingw32 -lgcc_s -lgcc -lmoldname -lmingwex -lmsvcrt -lpthread -ladvapi32 -lshell32 -luser32 -lkernel32 -lmingw32 -lgcc_s -lgcc -lmoldname -lmingwex -lmsvcrt /usr/lib/gcc/i686-w64-mingw32/9.3-posix/crtend.o